src/interpreter/crypto.rs

```diff
@@ -1,37 +1,35 @@
 use super::Interpreter;
-use crate::consts::{MEM_MAX_ACCESS_SIZE, VM_MAX_RAM};
-use crate::crypto;
+use crate::consts::{MEM_MAX_ACCESS_SIZE, MIN_VM_MAX_RAM_USIZE_MAX, VM_MAX_RAM};
 use crate::error::RuntimeError;
 
+use crate::arith::checked_add_word;
 use fuel_asm::PanicReason;
-use fuel_crypto::Hasher;
+use fuel_asm::PanicReason::ArithmeticOverflow;
+use fuel_crypto::{Hasher, Message, PublicKey, Signature};
 use fuel_types::{Bytes32, Bytes64, Word};
 
 impl<S> Interpreter<S> {
     pub(crate) fn ecrecover(&mut self, a: Word, b: Word, c: Word) -> Result<(), RuntimeError> {
-        if a > VM_MAX_RAM - Bytes64::LEN as Word
-            || b > VM_MAX_RAM - Bytes64::LEN as Word
-            || c > VM_MAX_RAM - Bytes32::LEN as Word
-        {
+        let bx = checked_add_word(b, Bytes64::LEN as Word)?;
+        let cx = checked_add_word(c, Bytes32::LEN as Word)?;
+
+        if a > VM_MAX_RAM - Bytes64::LEN as Word || bx > MIN_VM_MAX_RAM_USIZE_MAX || cx > MIN_VM_MAX_RAM_USIZE_MAX {
             return Err(PanicReason::MemoryOverflow.into());
         }
 
-        let (a, b, c) = (a as usize, b as usize, c as usize);
-
-        let bx = b + Bytes64::LEN;
-        let cx = c + Bytes32::LEN;
+        let (a, b, bx, c, cx) = (a as usize, b as usize, bx as usize, c as usize, cx as usize);
 
-        let e = &self.memory[c..cx];
-        let sig = &self.memory[b..bx];
+        // Safety: memory bounds are checked
+        let signature = unsafe { Signature::as_ref_unchecked(&self.memory[b..bx]) };
+        let message = unsafe { Message::as_ref_unchecked(&self.memory[c..cx]) };
 
-        match crypto::secp256k1_sign_compact_recover(sig, e) {
-            Ok(pk) => {
-                self.try_mem_write(a, pk.as_ref())?;
+        match signature.recover(message) {
+            Ok(public) => {
```

Comment:
nit: `public` is not the clearest name, only from previous `pk` name was able to decipher what it is
maybe: `pk`, `pub_key`, `public_key` ?

Reply:
absolutely, will change

Reply:
fixed